File: kraken/src/main/java/glitch/kraken/services/request/TopVideosRequest.java
package glitch.kraken.services.request;

import glitch.api.http.HttpClient;
import glitch.api.http.HttpRequest;
import glitch.api.http.Routes;
import glitch.api.objects.enums.VideoType;
import glitch.kraken.object.enums.VideoPeriod;
import glitch.kraken.object.enums.VideoSort;
import glitch.kraken.object.json.Game;
import glitch.kraken.object.json.Video;
import glitch.kraken.object.json.collections.Videos;
import glitch.service.AbstractRequest;
import java.util.*;
import java.util.stream.Collectors;
import reactor.core.publisher.Mono;

public class TopVideosRequest extends AbstractRequest<Video, Videos> {
    private final Set<VideoType> videoType = new LinkedHashSet<>();
    private final Set<Locale> language = new LinkedHashSet<>();
    private Integer limit;
    private Integer offset;
    private Game game;
    private VideoPeriod period;
    private VideoSort sort;

    public TopVideosRequest(HttpClient httpClient) {
        super(httpClient, Routes.get("/videos/top").newRequest());
    }

    public TopVideosRequest addVideoType(VideoType... types) {
        return addVideoType(Arrays.asList(types));
    }

    public TopVideosRequest addVideoType(Collection<VideoType> types) {
        videoType.addAll(types);
        return this;
    }

    public TopVideosRequest addLanguage(String... languages) {
        return addLanguage(Arrays.stream(languages).map(Locale::forLanguageTag).collect(Collectors.toSet()));
    }

    public TopVideosRequest addLanguage(Locale... languages) {
        return addLanguage(Arrays.asList(languages));
    }

    public TopVideosRequest addLanguage(Collection<Locale> languages) {
        language.addAll(languages);
        return this;
    }

    public TopVideosRequest setLimit(Integer limit) {
        this.limit = limit;
        return this;
    }

    public TopVideosRequest setOffset(Integer offset) {
        this.offset = offset;
        return this;
    }

    public TopVideosRequest setGame(Game game) {
        this.game = game;
        return this;
    }

    public TopVideosRequest setPeriod(VideoPeriod period) {
        this.period = period;
        return this;
    }

    public TopVideosRequest setSort(VideoSort sort) {
        this.sort = sort;
        return this;
    }

    public Mono<Videos> get() {

        if (limit != null && limit > 0 && limit <= 100) {
            request.queryParam("limit", limit);
        }

        if (offset != null && offset >= 0) {
            request.queryParam("offset", offset);
        }

        if (game != null) {
            request.queryParam("game", game.getName());
        }

        if (period != null) {
            request.queryParam("period", period.name().toLowerCase());
        }

        if (!videoType.isEmpty()) {
            request.queryParam("broadcast_type", videoType.stream().map(Enum::name).collect(Collectors.joining(",")).toLowerCase());
        }

        if (!language.isEmpty()) {
            request.queryParam("language", language.stream().map(Locale::getLanguage).collect(Collectors.joining(",")).toLowerCase());
        }

        if (sort != null) {
            request.queryParam("sort", sort.name().toLowerCase());
        }

        return httpClient.exchangeAs(request, Videos.class);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/request/TopVideosRequest.kt
package io.glitchlib.v5.service.request

import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest
import io.glitchlib.model.VideoType
import io.glitchlib.v5.model.VideoPeriod
import io.glitchlib.v5.model.VideoSort
import io.glitchlib.v5.model.json.Game
import java.util.Locale

class TopVideosRequest internal constructor() : AbstractRequest {
    //    AbstractRequest<Video, Videos>(httpClient, Routes.Companion.get("/videos/top").newRequest()) {
    private val videoType = mutableSetOf<VideoType>()
    private val language = mutableSetOf<Locale>()
    private var limit: Int? = null
    private var offset: Int? = null
    private var game: Game? = null
    private var period: VideoPeriod? = null
    private var sort: VideoSort? = null

    fun addVideoType(vararg types: VideoType) = addVideoType(types.toSet())

    fun addVideoType(types: Collection<VideoType>) = apply {
        videoType.addAll(types)
    }

    fun addLanguage(vararg languages: String) = addLanguage(languages.map(Locale::forLanguageTag))

    fun addLanguage(vararg languages: Locale) = addLanguage(languages.toList())

    fun addLanguage(languages: Collection<Locale>) = apply {
        language.addAll(languages)
    }

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun offset(offset: Int) = apply {
        this.offset = offset
    }

    fun game(game: Game) = apply {
        this.game = game
    }

    fun period(period: VideoPeriod) = apply {
        this.period = period
    }

    fun sort(sort: VideoSort) = apply {
        this.sort = sort
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (limit != null && limit!! > 0 && limit!! <= 100) {
            addQueryParameters("limit", limit!!.toString())
        }

        if (offset != null && offset!! >= 0) {
            addQueryParameters("offset", offset!!.toString())
        }

        if (game != null) {
            addQueryParameters("game", game!!.name)
        }

        if (period != null) {
            addQueryParameters("period", period!!.name.toLowerCase())
        }

        if (videoType.isNotEmpty()) {
            addQueryParameters("broadcast_type", videoType.joinToString(",") { it.name }.toLowerCase())
        }

        if (language.isNotEmpty()) {
            addQueryParameters("language", language.joinToString(",") { it.language }.toLowerCase())
        }

        if (sort != null) {
            addQueryParameters("sort", sort!!.name.toLowerCase())
        }
    }
}
